numalogic/udfs/trainer/_base.py

```diff
@@ -180,8 +180,22 @@ def exec(self, keys: list[str], datum: Datum) -> Messages:
         # Fetch data
         df = self.fetch_data(payload)
 
+        # Retry the training if df is returning None due to some errors/exception
+        # while fetching the data
+        if df is None:
```

Comment:
Instead of checking with None, could you return err in fetch_data and check with if err, like discussed yesterday?

Reply:
Are you implying the return value as a tuple of `(df, err)` ?

We had an offline discussion, underscoring that this approach aligns more with Golang conventions, rather than Python's. In this case, we can simply raise the error and capture the correct error class to accomplish what's necessary.

Reply:
yeah that should be fine too

Reply:
I dont see any changes here, can you raise and catch the error like Avik mentioned?

Reply:
We are catching it where the fetch function is called. We donot catch it here.

Reply:
that is fine, but checking with None here doesn't seem correct to me. Don't client calls return None in any scenario?

Reply:
discussed it offline, we agreed None should be fine for now. But, we would need look into this again sometime later, to see how it behaves during exceptions like connection errors. @s0nicboOm